### PR TITLE
Collect http.client_ip

### PR DESCRIPTION
**Pull request by Anilm3**
### Description

Collect `http.client_ip` using the resolution algorithm defined in the relevant RFC.

### Readiness checklist
- [ ] (only for Members) Changelog has been added to the release document.
- [ ] Tests added for this feature/bug.

### Reviewer checklist
- [ ] Appropriate labels assigned.
- [ ] Milestone is set.
- [ ] Changelog has been added to the release document. For community contributors the reviewer is in charge of this task.

### CODE REVIEW

File: ext/php7/ip_extraction.c [hunk 1]
```diff
@@ -0,0 +1,557 @@
...
+            .mask.s6_addr = {0xFE}, // /7
+        },
+    };
+    // clang-format on
+
+    unsigned __int128 addr_i;
+    memcpy(&addr_i, addr->s6_addr, sizeof(addr_i));
+
+    for (unsigned i = 0; i < ARRAY_SIZE(priv_ranges); i++) {
+        __auto_type range = &priv_ranges[i];
```
Comment (bwoebi) [inline ext/php7/ip_extraction.c:544]: Can we avoid these GNU extensions here (__int128 and __auto_type)? If we ever want to port this to Windows: msvc does not support these.

File: ext/php7/serializer.c [hunk 2]
```diff
@@ -570,6 +571,10 @@ void ddtrace_set_root_span_properties(ddtrace_span_t *span) {
         }
     }
 
+    if (!get_DD_TRACE_CLIENT_IP_HEADER_DISABLED()) {
+        ddtrace_extract_ip_from_headers(&PG(http_globals)[TRACK_VARS_SERVER], meta);
```
Comment (bwoebi) [inline ext/php7/serializer.c:575]: This needs a guard whether this is indeed an array - there's nothing preventing the user from redefining these...

**Review by bwoebi — COMMENTED** [review]
Generally we prefix private static symbols with `dd_` instead of just `_`.

File: ext/php7/ip_extraction.c [hunk 3]
```diff
@@ -0,0 +1,557 @@
...
+        (header_map_node){zend_string_init_interned(ZEND_STRL("HTTP_TRUE_CLIENT_IP"), 1),
+                          zend_string_init_interned(ZEND_STRL("true-client-ip"), 1), &_parse_plain};
+
+    _remote_addr_key = zend_string_init_interned(ZEND_STRL("REMOTE_ADDR"), 1);
+}
+
+static void _update_ipheader(zend_string *value) {
+    if (_ipheader) {
+        if (ZSTR_LEN(value) == ZSTR_LEN(_ipheader) &&
+            memcmp(ZSTR_VAL(value), ZSTR_VAL(_ipheader), ZSTR_LEN(value)) == 0) {
```
[the remark this answers is not in the document]
Comment (Anilm3) [reply ext/php7/ip_extraction.c:92]: a) Will do.
b) Indeed you're right, last minute change without thinking... I'll need to store the raw value.

File: ext/php7/ip_extraction.c [hunk 4]
```diff
@@ -0,0 +1,557 @@
...
+    MAX_HEADER_ID
+} header_id;
+
+static header_map_node header_map[MAX_HEADER_ID];
+static zend_string *_remote_addr_key;
+#if ZTS
+static __thread zend_string *_ipheader;
+#else
+static zend_string *_ipheader;
+#endif
```
[the remark this answers is not in the document]
Comment (Anilm3) [reply ext/php7/ip_extraction.c:53]: The alternative is not caching and just generating this on every call to `ddtrace_extract_ip_from_headers` which is also fine, I think.

File: ext/php7/ip_extraction.c [hunk 5]
```diff
@@ -0,0 +1,557 @@
...
+        end_cur = _skip_non_ws(p, end_cur);
+
+        succ = _parse_ip_address_maybe_port_pair(p, end_cur - p, out);
+        if (succ) {
+            succ = !_is_private(out);
+            if (succ) {
+                return out;
+            }
+        }
+    try_next:
```
[the remark this answers is not in the document]
Comment (Anilm3) [reply ext/php7/ip_extraction.c:397]: Yep, we only care about IPs here.

File: ext/php7/ip_extraction.c [hunk 6]
```diff
@@ -0,0 +1,557 @@
...
+        bool succ;
+        succ = _parse_forwarded(value, &out);
+        if (succ) {
+            return _ipaddr_to_zstr(&out);
+        }
+
+        succ = _parse_x_forwarded_for(value, &out);
+        if (succ) {
+            return _ipaddr_to_zstr(&out);
+        }
```
[the remark this answers is not in the document]
Comment (bwoebi) [reply ext/php7/ip_extraction.c:207]: Oh, you are right :-)

File: ext/php7/ip_extraction.c (hunk 1 above)
Comment (Anilm3) [reply ext/php7/ip_extraction.c]: Done

File: ext/php7/ip_extraction.c [hunk 7]
```diff
@@ -0,0 +1,557 @@
...
+static bool _parse_ip_address(const char *_addr, size_t addr_len, ipaddr *out);
+static bool _parse_ip_address_maybe_port_pair(const char *addr, size_t addr_len, ipaddr *out);
+
+static bool _parse_x_forwarded_for(zend_string *zvalue, ipaddr *out) {
+    const char *value = ZSTR_VAL(zvalue);
+    const char *end = value + ZSTR_LEN(zvalue);
+    bool succ;
+    do {
+        for (; value < end && *value == ' '; value++) {
+        }
```
[the remark this answers is not in the document]
Comment (Anilm3) [reply ext/php7/ip_extraction.c]: Done

File: ext/php7/ip_extraction.c (hunk 4 above)
[the remark this answers is not in the document]
Comment (Anilm3) [reply ext/php7/ip_extraction.c]: Not caching any longer.

**Review by bwoebi — APPROVED** [review]
Looks good to me now :-)

While I'm still not too happy with some logic duplication between serializer and ip extractor, we always can improve another time.

**Pull request merged**